src/zc/intid/tests.py

```diff
@@ -308,22 +366,198 @@ class I(zope.interface.Interface):
+    if setsite:
+        setSite(folder)
+    return folder.getSiteManager()
+
+class KeyReferenceStub(object):
+
+    def __init__(self, obj):
+        self.obj = obj
+
+class ReferenceSetupMixin(object):
+    """Registers adapters ILocation->IConnection and IPersistent->IReference"""
+
+    def setUp(self):
+        testing.setUp()
+        eventtesting.setUp()
+        traversingSetUp()
+        setHooks()
+        provideAdapter(SiteManagerAdapter, (Interface,), IComponentLookup)
+
+        self.root = rootFolder()
+        createSiteManager(self.root, setsite=True)
+
+        provideAdapter(
+            KeyReferenceStub, (IPersistent, ), IKeyReference)
+
+    def tearDown(self):
+        resetHooks()
+        setSite()
+        testing.tearDown()
+        self.assertIs(getSiteManager(), getGlobalSiteManager())
+
+
+class TestSubscribers(ReferenceSetupMixin, unittest.TestCase):
+
+    __parent__ = None
+    __name__ = None
+
+    def setUp(self):
+        from zope.site.folder import Folder
+
+        ReferenceSetupMixin.setUp(self)
+
+        sm = getSiteManager(self.root)
+        self.utility = IntIds("iid")
+        sm.registerUtility(self.utility, name='1', provided=IIntIds)
+
+        self.root['folder1'] = Folder()
+        self.root['folder1']['folder1_1'] = self.folder1_1 = Folder()
+        self.root['folder1']['folder1_1']['folder1_1_1'] = Folder()
+
+        sm1_1 = createSiteManager(self.folder1_1)
+        self.utility1 = IntIds("liid")
+        sm1_1.registerUtility(self.utility1, name='2', provided=IIntIds)
+
+        provideHandler(intIdEventNotify, (IIdEvent,))
+        provideHandler(intIdEventNotify, (IIntIdEvent,))
+
+        self.raw_events = []
+        self.obj_events = []
+
+        def obj_event(*args):
+            self.obj_events.append(args)
+
+        provideHandler(self.raw_events.append, [IIntIdEvent])
+        provideHandler(self.raw_events.append, [IIdEvent])
+
+        provideHandler(obj_event, [IFolder, IIntIdEvent])
+        provideHandler(obj_event, [IFolder, IIdEvent])
+
+        provideHandler(self.raw_events.append, [ISubscriberEvent])
+
+    def test_no_KeyReference(self):
+        # Nothing happens for something that can't be a KeyReference
+        addIntIdSubscriber(self, ObjectAddedEvent(self))
+        removeIntIdSubscriber(self, ObjectRemovedEvent(self))
+
+        self.assertEqual([], self.raw_events)
+        self.assertEqual([], self.obj_events)
+
+    def test_removeIntIdSubscriber(self):
+        parent_folder = self.root['folder1']['folder1_1']
+        folder = self.root['folder1']['folder1_1']['folder1_1_1']
+        id = self.utility.register(folder)
+        id1 = self.utility1.register(folder)
+        self.assertEqual(self.utility.getObject(id), folder)
+        self.assertEqual(self.utility1.getObject(id1), folder)
+        setSite(self.folder1_1)
+
+        events = self.raw_events
+        objevents = self.obj_events
+
+        del events[:]
+        del objevents[:]
```

Comment:
It wasn't necessary in `test_addIntIdSubscriber`, but it was needed at that location in `test_removeIntIdSubscriber` because the manual registration just above generated  events.

I've simplified all of this.